ipa-core/src/protocol/context/dzkp_field.rs

```diff
@@ -88,6 +90,142 @@ impl<const P: usize> FromIterator<Fp61BitPrime> for Vec<[Fp61BitPrime; P]> {
     }
 }
 
+/// Construct indices for the `convert_values` lookup tables.
+///
+/// `b0` has the least significant bit of each index, and `b1` and `b2` the subsequent
+/// bits. This routine rearranges the bits so that there is one table index in each
+/// nibble of the output. The technique is similar to the matrix transposes in
+/// `secret_sharing::vector::transpose`, which are based on a technique in chapter 7
+/// of "Hacker's Delight" (2nd ed.).
+///
+/// Output word `j` contains the table indices for input positions `i` having (i%4) == j.
+#[must_use]
+fn convert_values_table_indices(b0: u128, b1: u128, b2: u128) -> [u128; 4] {
+    let const_55 = u128::from_le_bytes([0x55; 16]);
+    let const_aa = u128::from_le_bytes([0xaa; 16]);
+
+    let const_33 = u128::from_le_bytes([0x33; 16]);
+    let const_cc = u128::from_le_bytes([0xcc; 16]);
+
+    // `k` = keep, `s` = swap
+    let b0k = b0 & const_55;
+    let b0s = b0 & const_aa;
+    let b1s = b1 & const_55;
+    let b1k = b1 & const_aa;
+    let b2k = b2 & const_55;
+    let b2s = b2 & const_aa;
+
+    // Swap the bits having (i%2) == 1 in b0/b2 with the bits having (i%2) == 0 in
+    // b1/b3, so that `c0` and `c1` have the lower halves of each index, and `c2` and
+    // `c3` have the upper halves of each index. (b3 is implicitly zero.)
+    let c0 = b0k | (b1s << 1); // 0s and 2s
+    let c1 = b1k | (b0s >> 1); // 1s and 3s
+    let c2 = b2k;
+    let c3 = b2s >> 1;
+
+    let c0k = c0 & const_33; // 0s
+    let c0s = c0 & const_cc; // 2s
+    let c1k = c1 & const_33; // 1s
+    let c1s = c1 & const_cc; // 3s
+    let c2s = c2 & const_33;
+    let c2k = c2 & const_cc;
+    let c3s = c3 & const_33;
+    let c3k = c3 & const_cc;
+
+    // Swap the bits having (i%4) ∈ {2,3} in c0/c1 with the bits having (i%4) ∈ {0,1} in
+    // c2/c3.
+    let y0 = c0k | (c2s << 2); // 0s
+    let y1 = c1k | (c3s << 2); // 1s
+    let y2 = c2k | (c0s >> 2); // 2s
+    let y3 = c3k | (c1s >> 2); // 3s
+
+    [y0, y1, y2, y3]
+}
+
+static TABLE_RIGHT: LazyLock<[[Fp61BitPrime; 4]; 8]> = LazyLock::new(|| {
```

Comment:
Isn't `LazyLock` basically free after the first use where it gets initialized? Unless I'm wrong about that, it doesn't seem worth it to me.

Reply:
my understanding that it requires an atomic load on every get, otherwise I don't understand how it would work. In this code this may actually be suboptimal as we will have all cores hitting that load and potentially thrashing their L1 cache line.

Reply:
I created a benchmark. On a version that uses `static mut` instead of `LazyLock`, any performance difference is smaller than the run-to-run noise (which is a bit larger than the criterion output would suggest, I get timings from 1.32 to 1.36 µs -- but other than not doing anything compute-intensive in parallel, I haven't taken steps to eliminate background noise on my machine).

```
Main:

dzkp_convert_values/convert
                        time:   [7.6663 µs 7.6706 µs 7.6752 µs]

This PR:

dzkp_convert_values/convert
                        time:   [1.3499 µs 1.3516 µs 1.3534 µs]
```